packages/flutter/lib/src/services/text_editing_delta.dart

```diff
@@ -0,0 +1,448 @@
+      );
+    } else if ((replacementDestinationStart == replacementDestinationEnd || isInsertingInsideComposingRegion) &&
+        !isOriginalComposingRegionTextChanged) {  // Insertion.
+      return TextEditingDeltaInsertion(
+        oldText: oldText,
+        deltaText: replacementSource.substring(replacementDestinationEnd - replacementDestinationStart, (replacementDestinationEnd - replacementDestinationStart) + (replacementSource.length - (replacementDestinationEnd - replacementDestinationStart))),
+        deltaRange: TextRange(
+          start: replacementDestinationEnd,
+          end: replacementDestinationEnd,
+        ),
+        selection: newSelection,
+        composing: newComposing,
+      );
+    } else if (isReplaced) {  // Replacement.
+      return TextEditingDeltaReplacement(
+        oldText: oldText,
+        deltaText: replacementSource,
+        deltaRange: TextRange(
+          start: replacementDestinationStart,
+          end: replacementDestinationEnd,
+        ),
+        selection: newSelection,
+        composing: newComposing,
+      );
+    }
+    assert(false);
+    return TextEditingDeltaNonTextUpdate(
+      oldText: oldText,
+      selection: newSelection,
+      composing: newComposing,
+    );
+  }
+
+  /// The old text state before the delta has occurred.
+  final String oldText;
+
+  /// The value represents the text that is being inserted/deleted by this delta.
+  ///
+  /// This value will slightly vary based on the [TextEditingDeltaType]:
+  ///
+  ///  * For a [TextEditingDeltaType.insertion] this will be the character/s being
+  /// inserted.
+  ///
+  ///  * For a [TextEditingDeltaType.deletion] this will be an empty string.
+  ///
+  ///  * For a [TextEditingDeltaType.replacement] this will be the text that is
+  /// replacing the [TextEditingDelta.deltaRange].
+  ///
+  ///  * For a [TextEditingDeltaType.nonTextUpdate] this will be an empty string.
+  final String deltaText;
+
+  /// This value can either represent a range of text that the delta is changing
+  /// or if is a collapsed range then it represents the point where this delta
+  /// began.
+  ///
+  /// This value will slightly vary based on the [TextEditingDeltaType]:
+  ///
+  ///  * For a [TextEditingDeltaType.insertion] this will be a collapsed range
+  /// representing the cursor position where the insertion began.
+  ///
+  ///  * For a [TextEditingDeltaType.deletion] this will be the range of text
+  /// that was deleted.
+  ///
+  ///  * For a [TextEditingDeltaType.replacement] this will be the range of
+  /// characters that are being replaced.
+  ///
+  ///  * For a [TextEditingDeltaType.nonTextUpdate] this will be a collapsed range
+  /// of (-1,-1).
+  final TextRange deltaRange;
+
+  /// {@template flutter.services.TextEditingDelta.deltaType}
+  /// The type of delta that has occured.
+  ///
+  /// See [TextEditingDeltaType] for more information.
+  /// {@endtemplate}
+  TextEditingDeltaType get deltaType;
```

Comment:
The reason I kept the `deltaType` is because doing this type of dynamic check using `is` is discouraged in the style guide. Having `deltaType` let's us avoid this pattern, but if this is a common pattern i'm alright with removing it.

Reply:
I couldn't find that in the style guide, could you point me to the section that talks about avoiding doing runtime type checks?

Reply:
https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-interleaving-multiple-concepts-together, https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#tldr  

Sure I was referring to these sections. Where it uses the example of checking for types of widgets using `is`, and says to avoid `is`. Could this also be applicable in this case?

Reply:
Ah I see. But in this case you actually have a __fixed__ set of subtypes (i.e., these delta types are sealed classes ) so IMO this is a bit different from what the style guide is trying to discourage (APIs that have special handling for some certain classes). But maybe there's something evil about `is` itself. /cc @Hixie

Reply:
After updating the classes to not have `deltaText` and `deltaRange`, but more explicit `textInserted` and `insertionOffset` having a `deltaType` seems to make a lot less sense.